chemprop/train/metrics.py

```diff
@@ -213,6 +229,66 @@ def accuracy(targets: List[int], preds: Union[List[float], List[List[float]]], t
     return accuracy_score(targets, hard_preds)
 
 
+def recall_metric(targets: List[int], preds: Union[List[float], List[List[float]]], threshold: float = 0.5) -> float:
+    """
+    Computes the recall of a binary prediction task using a given threshold for generating hard predictions.
+
+    Alternatively, computes recall for a multiclass prediction task by picking the largest probability.
+
+    :param targets: A list of binary targets.
+    :param preds: A list of prediction probabilities.
+    :param threshold: The threshold above which a prediction is considered positive.
+    :return: The computed recall.
+    """
+    if type(preds[0]) == list:  # multiclass
+        hard_preds = [p.index(max(p)) for p in preds]
+    else:
+        hard_preds = [1 if p > threshold else 0 for p in preds]  # binary prediction
+
+    return recall_score(targets, hard_preds)
+
+
+def precision_metric(targets: List[int], preds: Union[List[float], List[List[float]]], threshold: float = 0.5) -> float:
+    """
+    Computes the precision of a binary prediction task using a given threshold for generating hard predictions.
+
+    Alternatively, computes precision for a multiclass prediction task by picking the largest probability.
+
+    :param targets: A list of binary targets.
+    :param preds: A list of prediction probabilities.
+    :param threshold: The threshold above which a prediction is considered positive.
+    :return: The computed precision.
+    """
+    if type(preds[0]) == list:  # multiclass
+        hard_preds = [p.index(max(p)) for p in preds]
+    else:
+        hard_preds = [1 if p > threshold else 0 for p in preds]  # binary prediction
```

Comment:
This part of function is repeated many times in different metric functions. It's better to define a separate function that calculates the `hard_preds` than copy and paste it.